rfcs/react-components/convergence/custom-styling.md

```diff
@@ -0,0 +1,476 @@
+
+## Problem
+
+Partners migrating from v0/v8 to v9 have brought up that v9's custom styling does not provide the same granularity of
+customization as previous versions. See the appendix for detailed partner use cases and scenarios.
+
+Partners producing component libraries or component design surfaces are the most concerned about the lack of
+customization. They would like to customize the style of a single type of component without changing the style of
+other components.
+
+Example: Change the border radius and border width of all Button components without changing those styles for
+Input components.
+
+## Option A: React.Context useCustomStyles hooks
+
+Each component could check context for a useCustomStyle hook within the useComponentStyle hook.
+If this hook was defined, it is called at the end of the method before returning state.
+
+> These example use Component as a substitute for a component like Button or Avatar.
+
+```tsx
+const customStyles = useContextSelector(ComponentCustomStylesContext, context => context?.useCustomStyles);
+
+//...component styling here
+
+return customStyles ? customStyles(state) : state;
+```
+
+Each component defines own types for props and state, so would need to define its own context value types,
+context, and context provider.
+
+```tsx
+export type ComponentCustomStylesContextValue = {
+  useCustomStyles?: (state: ComponentState) => ComponentState;
+};
+
+export const ComponentCustomStylesContext: Context<ComponentCustomStylesContextValue> = createContext<ComponentCustomStylesContextValue>(
+  {},
+);
+
+export const ComponentCustomStylesContextProvider = ComponentCustomStylesContext.Provider;
+```
+
+When a caller wanted to customize the component, they use the provider. Callers would be able to control custom
+styling scope just like Fluent Provider. While they would need to use a provider per custom styled component,
+these could be aggregated into a component that transcludes children.
+
+```tsx
+<App>
+  <FluentProvider theme={webLightTheme}>
+    <ComponentCustomStylesContextProvider value={{ useCustomStyles }}>
+      {/* other component hierarchy */}
+      <Component>Hello custom styles!</Component>
+    </ComponentCustomStylesContextProvider>
+  </FluentProvider>
+</App>
+```
+
+### 👍
+
+- Custom styling gets the same power as the default component useStyles hook.
+- This builds on using slots as the addressable items for applying styles.
+- The custom styles can be built using Griffel with build-time optimization.
+- Can be applied at different scope like FluentProvider.
+- Does not require composing or introducing a new component type that must be adopted in lieu of the default component.
+- Fewer CSS specificity problems as styles can get overwritten.
+- Provides the type-safety expected from Fluent UI React.
+- Minimal code to add to each component. Could be put into the code generation template.
+- Does not significantly affect rendering performance.
```

Comment:
I would measure it first: additional calls of `mergeClasses()` could be expensive if there is 1k `Button`s in a tree.

Reply:
As consumers customize, they would need to account for the increased cost of mergeClasses calls.  They could apply their class and the existing classes without mergeClasses if they want to avoid the overhead of the merge and take the potential duplication hit right?